components/promise-component.js

```diff
@@ -0,0 +1,66 @@
+// promise-component.js
+
+var _ = require('underscore');
+var noflo = require('noflo');
+
+module.exports = function(def){
+    if (_.isEmpty(def.inPorts)) throw Error("Missing inPorts");
+    if (def.resolvePort && !_.isString(def.resolvePort.name))
+        throw Error("Missing resolvePort name");
+    if (def.rejectPort && !_.isString(def.rejectPort.name))
+        throw Error("Missing rejectPort name");
+    var resolvePort = def.resolvePort || _.defaults({
+        name: 'out'
+    }, _.propertyOf(def.outPorts)('out')) || {
+        name: 'out',
+        description: "Resolved result of the updater",
+        datatype: 'object',
+        addressable: false,
+        buffered: false,
+        required: false,
+        caching: false
+    };
+    var rejectPort = def.rejectPort || _.defaults({
+        name: 'error'
+    }, _.propertyOf(def.outPorts)('error')) || {
+        name: 'error',
+        description: "Rejected error of the updater",
+        datatype: 'object',
+        addressable: false,
+        buffered: false,
+        required: false,
+        caching: false
+    };
+    var inPorts = _.mapObject(def.inPorts, function(port, name){
+        var process = port.process || function(event, payload) {
+            var key = 'on' + event;
+            if (port[key]) return port[key].call(this.nodeInstance, payload);
+        };
+        return _.defaults({
+            process: function(event, payload, socketIndex) {
+                var self = this;
+                var outPorts = this.nodeInstance.outPorts;
+                Promise.resolve().then(function(){
+                    return process.call(self, event, payload, socketIndex);
```

Comment:
Except perhaps for special components, I think we should be moving toward all ports using the same 'process' function.   Different logic needed by different wrappers (component types) should be moved into the wrapper functions.  See https://github.com/rdf-pipeline/framework/wiki/Wrapper-Interface  The logic specific to each wrapper (component type) should be in that wrapper's fRunupdater function.  We should discuss this.